src/WebJobs.Script/Host/FunctionMetadataManager.cs

```diff
@@ -18,36 +21,95 @@ namespace Microsoft.Azure.WebJobs.Script
     {
         private const string _functionConfigurationErrorMessage = "Unable to determine the primary function script.Make sure atleast one script file is present.Try renaming your entry point script to 'run' or alternatively you can specify the name of the entry point script explicitly by adding a 'scriptFile' property to your function metadata.";
-        private readonly bool _isHttpWorker;
-        private readonly Lazy<ImmutableArray<FunctionMetadata>> _functionMetadataArray;
-        private readonly IOptions<ScriptJobHostOptions> _scriptOptions;
+        private readonly IServiceProvider _serviceProvider;
         private readonly IFunctionMetadataProvider _functionMetadataProvider;
-        private readonly ILogger _logger;
+        private bool _isHttpWorker;
+        private bool _servicesReset = false;
+        private ILogger _logger;
+        private IOptions<ScriptJobHostOptions> _scriptOptions;
+        private ImmutableArray<FunctionMetadata> _functionMetadataFilteredArray;
+        private ImmutableArray<FunctionMetadata> _functionMetadataAllArray;
+        private IEnumerable<IFunctionProvider> _functionProviders;
         private Dictionary<string, ICollection<string>> _functionErrors = new Dictionary<string, ICollection<string>>();
 
-        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IOptions<HttpWorkerOptions> httpWorkerOptions, ILoggerFactory loggerFactory)
+        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IEnumerable<IFunctionProvider> functionProviders, IOptions<HttpWorkerOptions> httpWorkerOptions, IScriptHostManager scriptHostManager, ILoggerFactory loggerFactory)
         {
             _scriptOptions = scriptOptions;
+            _serviceProvider = scriptHostManager as IServiceProvider;
             _functionMetadataProvider = functionMetadataProvider;
             _logger = loggerFactory.CreateLogger(LogCategories.Startup);
-            _functionMetadataArray = new Lazy<ImmutableArray<FunctionMetadata>>(LoadFunctionMetadata);
-            _isHttpWorker = httpWorkerOptions.Value.Description != null;
-        }
+            _isHttpWorker = httpWorkerOptions?.Value?.Description != null;
+            _functionProviders = functionProviders;
 
-        public ImmutableArray<FunctionMetadata> Functions => _functionMetadataArray.Value;
+            scriptHostManager.HostInitializing += ResetScriptHostServices;
+        }
 
         public ImmutableDictionary<string, ImmutableArray<string>> Errors { get; private set; }
 
+        /// <summary>
+        /// Gets the function metadata array from all providers.
+        /// </summary>
+        /// <param name="forceRefresh">Forces reload from all providers.</param>
+        /// <param name="includeBlocked">Includes all functions, including blocked ones.</param>
```

Comment:
What is a "blocked" function?

Reply:
If `Functions` in host.json is specified for the list of enabled functions, the functions that are not present there is what I termed as "blocked" function. Is there a better name?

Reply:
Do you mean disabled functions?

Reply:
I see. It goes to your earlier question on whether we should include non-whitelisted in this response. Since the whole point of whitelisting is to exclude functions completely, I'm wondering if we shouldn't just remove this switch and never return non-whitelisted functions. An "excluded" function won't be loaded by the host, won't have a listener, it won't be running, and it shouldn't be shown in portal, be monitored by ScaleController, etc. I know you had identified a discrepency where the admin APIs weren't applying the whitelist and you're trying to preserve that. Excluded functions also shouldn't be returned as part of the sync triggers response. Is it currently?

Perhaps we take this opportunity to clean that up. The whitelist is rarely used by people, so likely even before your changes we're listing functions that we shouldn't be.

Reply:
@pragnagopa, these are for non-whitelisted functions.

Yes, today it is in fact returned as part of the sync triggers payload. Right now, I removed it from sync triggers payload but left it in "List Functions" to avoid breaking change. But with my recent understanding that "List Functions" uses the sync triggers cache, that will create a discrepancy.

@fabiocav mentioned earlier that we do have this `functions` (whitelist) property in the `host.json` schema as well as our documentation https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json#functions

Though, reading it, we specifically ask users to not do that in Azure.

@fabiocav had some concerns with breaking this behavior. I would love to hear his thoughts here too.

Reply:
Rename this param to applyWhitelist to match existing terminology